python/tvm/relax/op/distributed/distributed.py

```diff
@@ -59,3 +59,28 @@ def redistribute(input: Expr, device_mesh: DeviceMesh, placement: Placement) ->
       The tensor after redistribution.
     """
     return _ffi_api.redistribute(input, device_mesh, placement)  # type: ignore
+
+
+def redistribute_replica_to_shard(input: Expr, num_workers: int, axis: int) -> Expr:
```

Comment:
We do need runtime support. Currently Disco runtime treats num_workers as a constant, and legalize_ops of ccl op will throw away `num_workers` when converting to `call_dps_packed(ccl, ...)`.  This behavior needs to be changed if we want to make num_worker a `Expr`

Reply:
I would recommend we do a refactor on the whole stack to support dynamic num_workers if there is really a need in the future.

Reply:
That is correct, that we would require runtime support, but only if the `num_workers` is still dynamic after being lowered to either the disco runtime or the ccl op legalization.  It is easier to write a single dynamic implementation then specialize to a variety of static cases, than it is to write several distinct static implementations.  However, the initial writing of the dynamic implementation requires that it be expressible, even though it will be specialized out later in lowering.

Reply:
I don't understand the case you are talking about. Why do we want to use a symbolic var for a value that is constant after lowering? And why will we have a variety of static case to specialize, given disco runtime regards `num_workers` as a global constant?

Reply:
One example is if you are comparing how performance scales with the number of GPUs, then you need a way to specify the number of GPUs.  Each data point collected would be for a specialized value of `num_workers`.

Effectively, having `num_workers: Expr` de-couples the communication mechanics from the choice of how many workers to use.  It still can be explicitly specified at any stage of lowering, and it still must be statically-known after lowering, but we gain flexibility before then.  This has a number of benefits, for example:

* Predicted memory usage.  If the number of workers is stored symbolically, adding up the size of all live values at any point gives the memory footprint as a function of the number of workers.  Requiring the number of workers to be static at all points of lowering prevents this analysis.

* Consistent optimization.  If an optimization is applicable regardless of the number of workers, the optimization should be applied at a point when the number of workers is unknown.  This prevents a developer from accidentally making a less general optimization.  (e.g. By using a sharded tensor shape in the pattern-matching.

Reply:
got it. This makes sense. Since redistribute_R_to_S shares the attribute with scatter_from_worker0, I'd like to open up a followup PR for this.

Reply:
@Lunderberg I just come up with an additional question on the attrs->args change: Do we need to recompile each time for different num_workers after this change with specialization?